code-dot-org/code-dot-org
# Fix placeholder text translation

[FND-1359](https://codedotorg.atlassian.net/browse/FND-1359)
The original work at https://github.com/code-dot-org/code-dot-org/pull/36983 enabled translation for placeholder texts in Blockly blocks. However, it covers only 1 type of block which is `text` block. This PR adds support for 2 more types of block:  `studio_showTitleScreen` and `studio_ask`.


## Examples
**Example 1**: Translating 2 placeholder strings in a `studio_showTitleScreen` block from English to Vietnamese.
Original|Translated
--|--
<img width="163" alt="Screen Shot 2021-01-17 at 8 48 50 PM" src="https://user-images.githubusercontent.com/46507039/104877943-cf2ac400-590f-11eb-8c95-75d76abefc59.png">|<img width="193" alt="Screen Shot 2021-01-17 at 8 46 15 PM" src="https://user-images.githubusercontent.com/46507039/104877941-cf2ac400-590f-11eb-8fa8-e94ba587dabb.png">

**Example 2**: Translating placeholder strings in a `studio_ask` block ("ask... to set...") and a `text` block ("...") from English to Vietnamese.
Original|Translated
--|--
<img width="569" alt="Screen Shot 2021-01-17 at 8 50 15 PM" src="https://user-images.githubusercontent.com/46507039/104877944-cfc35a80-590f-11eb-83f3-27843e741596.png">|<img width="416" alt="Screen Shot 2021-01-17 at 8 45 49 PM" src="https://user-images.githubusercontent.com/46507039/104877937-cdf99700-590f-11eb-9959-42b6d50eda22.png">


## Testing story
- Unit test `bundle exec rails test test/models/blockly_test.rb`.
- Local test:
  - Run `sync-in.rb` to extract new placeholder strings to `i18n/locales/source/course_content/` folder.
  - Fake the results of `sync-down.rb` by adding placeholder string translations to files in `i18n/locales/<locale>/course_content/` folder.
  - Run `sync-out.rb` to distribute the changes from `i18n/locales/<locale>/course_content/*` to `/dashboard/config/locales/placeholder_texts.<locale>.json`.
  - Open levels with blocks that contain placeholder strings and verify that placeholder strings are translated. Example levels are:
    - http://localhost-studio.code.org:3000/s/coursee-2020/stage/9/puzzle/1/lang/vi
    - http://localhost-studio.code.org:3000/s/coursee-2017/stage/20/puzzle/1/lang/vi?no_redirect=true
    - http://localhost-studio.code.org:3000/s/express-2017/stage/31/puzzle/4/lang/vi?no_redirect=true
